File: src/main/java/io/polygloat/model/Language.java
package io.polygloat.model;

import io.polygloat.dtos.request.LanguageDTO;

import javax.persistence.*;
import java.util.Set;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"repository_id", "name"}, name = "language_repository_name"),
        @UniqueConstraint(columnNames = {"repository_id", "abbreviation"}, name = "language_abbreviation_name")
},
        indexes = {
                @Index(columnList = "abbreviation", name = "index_abbreviation"),
                @Index(columnList = "abbreviation, repository_id", name = "index_abbreviation_repository")
        }
)
public class Language extends AuditModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "language")
    private Set<Translation> translations;

    @ManyToOne
    private Repository repository;

    private String abbreviation;

    private String name;

    public static Language fromRequestDTO(LanguageDTO dto) {
        Language language = new Language();
        language.setName(dto.getName());
        language.setAbbreviation(dto.getAbbreviation());
        return language;
    }

    public void updateByDTO(LanguageDTO dto) {
        this.name = dto.getName();
        this.abbreviation = dto.getAbbreviation();
    }

    public Repository getRepository() {
        return repository;
    }

    public String getName() {
        return name;
    }

    public Long getId() {
        return id;
    }

    public String getAbbreviation() {
        return abbreviation;
    }

    public String toString() {
        return "Language(id=" + this.getId() + ", abbreviation=" + this.getAbbreviation() + ", name=" + this.getName() + ")";
    }

    public Set<Translation> getTranslations() {
        return this.translations;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public void setTranslations(Set<Translation> translations) {
        this.translations = translations;
    }

    public void setRepository(Repository repository) {
        this.repository = repository;
    }

    public void setAbbreviation(String abbreviation) {
        this.abbreviation = abbreviation;
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: backend/data/src/main/kotlin/io/tolgee/model/Language.kt
package io.tolgee.model

import io.tolgee.activity.annotation.ActivityDescribingProp
import io.tolgee.activity.annotation.ActivityLoggedEntity
import io.tolgee.activity.annotation.ActivityLoggedProp
import io.tolgee.activity.annotation.ActivityReturnsExistence
import io.tolgee.dtos.request.LanguageRequest
import io.tolgee.events.OnLanguagePrePersist
import io.tolgee.events.OnLanguagePreRemove
import io.tolgee.model.mtServiceConfig.MtServiceConfig
import io.tolgee.model.translation.Translation
import io.tolgee.service.dataImport.ImportService
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.FetchType
import jakarta.persistence.Index
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.OneToOne
import jakarta.persistence.PrePersist
import jakarta.persistence.PreRemove
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.Size
import org.springframework.beans.factory.ObjectFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Configurable
import org.springframework.context.ApplicationEventPublisher
import org.springframework.transaction.annotation.Transactional

@Entity
@EntityListeners(Language.Companion.LanguageListeners::class)
@Table(
  uniqueConstraints = [
    UniqueConstraint(
      columnNames = ["project_id", "name"],
      name = "language_project_name",
    ),
    UniqueConstraint(
      columnNames = ["project_id", "tag"],
      name = "language_tag_name",
    ),
  ],
  indexes = [
    Index(
      columnList = "tag",
      name = "index_tag",
    ),
    Index(
      columnList = "tag, project_id",
      name = "index_tag_project",
    ),
  ],
)
@ActivityLoggedEntity
@ActivityReturnsExistence
class Language : StandardAuditModel(), ILanguage {
  @OneToMany(fetch = FetchType.LAZY, mappedBy = "language", orphanRemoval = true)
  var translations: MutableSet<Translation>? = null

  @ManyToOne(fetch = FetchType.LAZY)
  lateinit var project: Project

  @Column(nullable = false)
  @field:NotEmpty
  @ActivityLoggedProp
  @ActivityDescribingProp
  override var tag: String = ""

  @ActivityLoggedProp
  @ActivityDescribingProp
  @field:NotBlank
  override var name: String = ""

  @ActivityLoggedProp
  override var originalName: String? = null

  @field:Size(max = 20)
  @Column(length = 20)
  @ActivityLoggedProp
  @ActivityDescribingProp
  override var flagEmoji: String? = null

  @OneToOne(mappedBy = "targetLanguage", orphanRemoval = true, fetch = FetchType.LAZY)
  var mtServiceConfig: MtServiceConfig? = null

  @OneToOne(mappedBy = "targetLanguage", orphanRemoval = true, fetch = FetchType.LAZY)
  var autoTranslationConfig: AutoTranslationConfig? = null

  @OneToOne(mappedBy = "language", orphanRemoval = true, fetch = FetchType.LAZY)
  var stats: LanguageStats? = null

  @field:Size(max = 5000)
  @ActivityLoggedProp
  override var aiTranslatorPromptDescription: String? = null

  fun updateByDTO(dto: LanguageRequest) {
    name = dto.name
    tag = dto.tag
    originalName = dto.originalName
    flagEmoji = dto.flagEmoji
  }

  override fun toString(): String {
    return "Language(tag=$tag, name=$name, originalName=$originalName)"
  }

  companion object {
    @JvmStatic
    fun fromRequestDTO(dto: LanguageRequest): Language {
      val language = Language()
      language.name = dto.name
      language.tag = dto.tag
      language.originalName = dto.originalName
      language.flagEmoji = dto.flagEmoji
      return language
    }

    @Configurable
    class LanguageListeners {
      @Autowired
      lateinit var importServiceProvider: ObjectFactory<ImportService>

      @Autowired
      lateinit var eventPublisherProvider: ObjectFactory<ApplicationEventPublisher>

      @PrePersist
      fun prePersist(language: Language) {
        eventPublisherProvider.`object`.publishEvent(OnLanguagePrePersist(source = this, language))
      }

      @PreRemove
      @Transactional
      fun preRemove(language: Language) {
        importServiceProvider.`object`.onExistingLanguageRemoved(language)
        eventPublisherProvider.`object`.publishEvent(OnLanguagePreRemove(source = this, language))
      }
    }
  }
}
